### PR TITLE
update to analyzer 1.1.0 apis

### PR DESCRIPTION
**Pull request by pq**
/cc @bwilkerson @scheglov

### CODE REVIEW

File: lib/src/ast.dart [hunk 1]
```diff
@@ -233,15 +233,15 @@ File? locatePubspecFile(CompilationUnit compilationUnit) {
   }
 
   final file = resourceProvider.getFile(fullName);
-  var folder = file.parent;
+  var folder = file.parent2;
 
   // Look for a pubspec.yaml file.
-  while (folder != null) {
+  while (!folder.isRoot) {
```
Comment (scheglov) [inline lib/src/ast.dart:239]: You could use `FolderExtension.withAncestors` here, e.g.
```dart
for (var current in file.parent2.withAncestors) {
  ...
}
```

File: lib/src/rules/avoid_returning_null_for_future.dart [hunk 2]
```diff
@@ -63,17 +63,17 @@ class _Visitor extends SimpleAstVisitor<void> {
     if (parent == null) return;
 
     DartType? returnType;
```
Comment (scheglov) [inline lib/src/rules/avoid_returning_null_for_future.dart:65]: The `declaredElement` must be always set (to `ExecutableElement`) in resolved units, so the return type is never `null`.

File: lib/src/rules/avoid_returning_null_for_future.dart [hunk 3]
```diff
@@ -63,17 +63,17 @@ class _Visitor extends SimpleAstVisitor<void> {
...
-    bool? isAsync;
+    bool isAsync;
     if (parent is FunctionExpression) {
       returnType = parent.declaredElement?.returnType;
-      isAsync = parent.body?.isAsynchronous;
+      isAsync = parent.body.isAsynchronous;
```
Comment (scheglov) [inline lib/src/rules/avoid_returning_null_for_future.dart:69]: With the change to `declaredElement` you could get `isAsynchronous` from `ExecutableElement`.

File: lib/src/rules/avoid_returning_null_for_void.dart [hunk 4]
```diff
@@ -89,14 +89,14 @@ class _Visitor extends SimpleAstVisitor<void> {
     bool? isAsync;
     if (parent is FunctionExpression) {
       type = parent.declaredElement?.returnType;
-      isAsync = parent.body?.isAsynchronous;
+      isAsync = parent.body.isAsynchronous;
```
Comment (scheglov) [inline lib/src/rules/avoid_returning_null_for_void.dart:92]: Looks like a dup of the above.
Consider adding a helper to get the `ExecutableElement` and use in both places to access the return type and `isAsynchronous`.

**Review by scheglov — APPROVED** [review]

File: lib/src/ast.dart (hunk 1 above)
Comment (pq) [reply lib/src/ast.dart:239]: Nice!

File: lib/src/rules/avoid_returning_null_for_future.dart (hunk 2 above)
Comment (pq) [reply lib/src/rules/avoid_returning_null_for_future.dart:65]: Are you suggesting `!`s below?:

    returnType = parent.declaredElement!.returnType;

**Comment by pq** [comment]
Landing and can follow up with more tweaks.

**Pull request merged**